### PR TITLE
Change to new Aventer Mesos URL

### LINKED ISSUE #4282: Docker in the Toil appliance container does not know about --gpus
@glennhickey says that you can't use the Docker client we ship in the Toil appliance with the newly-upstreamed `--gpus` option:

```
docker run -it --rm --gpus all  nvidia/cuda:11.4.3-runtime-ubuntu20.04 bash
unknown flag: --gpus
See 'docker run --help'.
```

We should update our Docker client to one that supports this option.

Related to #4199.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1247)
┆Issue Number: TOIL-1247

### LINKED ISSUE #4289: Aventer APT source is broken again
https://rpm.aventer.biz/Ubuntu moved their whole Focal tree inside https://rpm.aventer.biz/Ubuntu/focal for no apparent reason. The apt source we use can't handle this, so the Toil appliance build fails.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1249)
┆Issue Number: TOIL-1249

### PR DESCRIPTION
**Pull request by adamnovak**
This ought to fix #4289.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Docker build has been updated with new Aventer Mesos deb URL (fixes #4290)
 * `docker` binary in the container has been updated to that included in the Ubuntu repos (fixes #4282)
 * Singularity in the appliance has been updated to 3.10 which is >=3.9, for cgroups v2 support.
 * Base Ubuntu container image for the appliance has been updated to 22.04, which has a new enough libc for Debian's Singularity 3.10 debs.

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by adamnovak** [comment]
I'm also glomming a fix for #4282 onto here since I am messing about with deb dependencies anyway.

**Comment by adamnovak** [comment]
OK, I had to adjust https://github.com/vgteam/dind to do some cgroup surgery on cgroups v2 machines, to get all the processes in the container *out* of the container's root cgroup, and into another cgroup, to allow new containers under the container's root cgroup to be created, because processes and cgroups with limits on them can't exist in the same cgroup under cgroups v2.

I also had to upgrade Singularity to one that works with cgroups v2, which meant I had to upgrade the base Ubuntu image to ine with a new enough libc for the newer Singularity builds.

**Comment by adamnovak** [comment]
Now I have Singularity working if the Kubernetes pod is privileged , or if it [has an `unconfined` AppArmor profile applied](https://kubernetes.io/docs/tutorials/security/apparmor/), but if not I get:

```
cd
singularity build -s -F ./sandbox docker://ubuntu:20.04
'singularity' -d -v 'exec' '-w' '-u' '-B' '' '--pwd' '/mnt' './sandbox' ls -lah /
```

```
VERBOSE [U=0,P=532]        user_namespace_init()         Create user namespace
VERBOSE [U=0,P=532]        create_namespace()            Create user namespace
VERBOSE [U=65534,P=532]    init()                        Spawn master process
DEBUG   [U=65534,P=532]    setup_userns_mappings()       Write deny to setgroups file
DEBUG   [U=65534,P=532]    setup_userns_mappings()       Write to GID map
DEBUG   [U=65534,P=532]    setup_userns_mappings()       Write to UID map
DEBUG   [U=0,P=551]        set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=0,P=551]        create_namespace()            Create mount namespace
ERROR   [U=0,P=551]        shared_mount_namespace_init() Failed to set mount propagation: Permission denied
```

This is related to https://github.com/apptainer/singularity/issues/5857 and to https://gitlab.nrp-nautilus.io/prp/nautilus-cluster/-/issues/570

The default Ubuntu AppArmor profiles on the new host nodes break Singularity (at least on cgroups v2). I think I have to turn apparmor off on the nodes because we can't just annotate all the pods. I don't think Fedora has AppArmor on?

**Comment by adamnovak** [comment]
I've modified the ASG templates for the cluster, but the nodes need to scale away and come back again before Singularity will work again, I think.

File: src/toil/lib/threading.py [hunk 1]
```diff
@@ -105,42 +105,77 @@ def cpu_count() -> Any:
...
         # We can't actually read these cgroup fields. Maybe we are a mac or something.
         logger.debug('Could not inspect cgroup: %s', traceback.format_exc())
-        cgroup_size = float('inf') # type: ignore
+
+    # CPU affinity may limit the size
+    affinity_size: Union[float, int] = float('inf')
+    if hasattr(os, 'sched_getaffinity'):
+        try:
+            logger.debug('CPU affinity available')
+            affinity_size = len(os.sched_getaffinity(0))
```
Comment (mr-c) [inline src/toil/lib/threading.py:169]: Consider using `len(psutil.Process().cpu_affinity()` which is supported for Linux, Windows, FreeBSD  (and we already use `psutil`)

https://psutil.readthedocs.io/en/latest/index.html#psutil.Process.cpu_affinity

Comment (mr-c) [reply src/toil/lib/threading.py:169]: Ah, nevermind, the implementation is quite similar between the two

https://github.com/python/cpython/blob/0076ca48e9cd116bb9d283d014fe69a72c9f1f6f/Modules/posixmodule.c#L7229
https://github.com/giampaolo/psutil/blob/5b3016ba3d7429c827c992699f50c2b7ff29b023/psutil/_psutil_linux.c#L239

File: docker/Dockerfile.py [hunk 2]
```diff
@@ -83,7 +84,7 @@ def heredoc(s):
 motd = ''.join(l + '\\n\\\n' for l in motd.splitlines())
 
 print(heredoc('''
-    FROM ubuntu:20.04
+    FROM ubuntu:22.04
```
Comment (DailyDreaming) [inline docker/Dockerfile.py:87]: Nice...

**Review by DailyDreaming — APPROVED** [review]
LGTM.  Thank you for this!

**Comment by adamnovak** [comment]
OK, using this and a general `mount` allow rule for AppArmor for Docker containers, I am able to run `kubectl run --limits='cpu=4000m,memory=4Gi,ephemeral-storage=16Gi' toil-test --rm -i --tty --overrides='{"spec": {"affinity": {"nodeAffinity": {"requiredDuringSchedulingIgnoredDuringExecution": {"nodeSelectorTerms": [{"matchExpressions": [{"key": "kubernetes.io/hostname", "operator": "NotIn", "values": ["ip-172-31-51-233.us-west-2.compute.internal"]}]}]}}}}}' --image quay.io/ucsc_cgl/toil:5.8.0a1-faf4ce788afa2937c987b88877c69143bf84f67d-py3.9 --command -- bash -c "cd ; wget https://github.com/common-workflow-language/cwl-v1.2/raw/v1.2.0/tests/tmap-job.json ; wget https://github.com/common-workflow-language/cwl-v1.2/raw/v1.2.0/tests/tmap-tool.cwl ; wget https://github.com/common-workflow-language/cwl-v1.2/raw/v1.2.0/tests/reads.fastq ; wget https://github.com/common-workflow-language/cwl-v1.2/raw/v1.2.0/tests/args.py ; export SINGULARITY_DOCKER_HUB_MIRROR=http://docker-registry.toil ; toil-cwl-runner --clean=always --logDebug --statusWait=10 --disableCaching --retryCount=0 --setEnv=SINGULARITY_DOCKER_HUB_MIRROR --batchSystem=single_machine --singularity --outdir=test-out tmap-tool.cwl tmap-job.json"` and show that I can run Singularity containers under Toil CWL workflows from within a Kubernetes pod on one of the new worker nodes.

Hopefully the Cactus and other integration tests are similarly happy with the move.

**Pull request merged**